contrib/pyzfs/libzfs/exceptions.py

```diff
@@ -0,0 +1,129 @@
+#
+# Copyright 2019 Hudson River Trading LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import errno
+
+
+class ZFSError(Exception):
+    errno = None
+    message = None
+    name = None
+
+    def __str__(self):
+        if self.name is not None:
+            return "[Errno %d] %s: '%s'" % (
+                self.errno, self.message, self.name)
+        else:
+            return "[Errno %d] %s" % (self.errno, self.message)
+
+    def __repr__(self):
+        return "%s(%r, %r)" % (
+            self.__class__.__name__, self.errno, self.message)
+
+
+class LibZfsInitError(ZFSError):
+    errno = -1
+    message = "Failed to open a handle to the ZFS driver"
+
+
+class ZpoolOpenError(ZFSError):
```

Comment:
For many of these "Fail to open..." error messages, it is very helpful to add the instance or name of the thing that failed. For example, many times we look for all pools and if one of them fails, we'd like to know which one. It should be easy to add the name or instance into these messages.

Reply:
Yup, when I raise this exception in certain spots, I do pass in the name of the pool that failed to open, etc.